include/wil/registry_helpers.h

```diff
@@ -0,0 +1,947 @@
+//*********************************************************
+//
+//    Copyright (c) Microsoft. All rights reserved.
+//    This code is licensed under the MIT License.
+//    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF
+//    ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
+//    TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+//    PARTICULAR PURPOSE AND NONINFRINGEMENT.
+//
+//*********************************************************
+#ifndef __WIL_REGISTRY_HELPERS_INCLUDED
+#define __WIL_REGISTRY_HELPERS_INCLUDED
+
+#if defined(_STRING_) || defined (_VECTOR_) || (defined (__cpp_lib_optional) && defined (_OPTIONAL_))
+#include <functional>
+#include <iterator>
+#endif
+
+#include <Windows.h>
+#include "resource.h"
+
+#ifdef _KERNEL_MODE
+#error This header is not supported in kernel-mode.
+#endif
+
+#ifndef __WIL_WINREG_
+#error This is required for ::wil::unique_hkey support
+#endif
+
+namespace wil
+{
+    namespace reg
+    {
+        enum class key_access
+        {
+            read,
+            readwrite,
+        };
+
+        namespace reg_view_details
+        {
+            constexpr DWORD get_value_flags_from_value_type(DWORD type) WI_NOEXCEPT
+            {
+                switch (type)
+                {
+                case REG_DWORD:
+                    return RRF_RT_REG_DWORD;
+                case REG_QWORD:
+                    return RRF_RT_REG_QWORD;
+                case REG_SZ:
+                    return RRF_RT_REG_SZ | RRF_RT_REG_EXPAND_SZ | RRF_NOEXPAND;
+                case REG_EXPAND_SZ:
+                    return RRF_RT_REG_SZ | RRF_RT_REG_EXPAND_SZ;
+                case REG_MULTI_SZ:
+                    return RRF_RT_REG_MULTI_SZ;
+                case REG_BINARY:
+                    return RRF_RT_REG_BINARY;
+                    // the caller can directly specify their own flags if they need to
+                default:
+                    return type;
+                }
+            }
+
+            constexpr DWORD get_access_flags(key_access access) WI_NOEXCEPT
+            {
+                switch (access)
+                {
+                case key_access::read:
+                    return KEY_READ;
+                case key_access::readwrite:
+                    return KEY_ALL_ACCESS;
+                }
+                FAIL_FAST();
+            }
+
+#if defined(_VECTOR_) && defined(_STRING_) && defined(WIL_ENABLE_EXCEPTIONS)
+            template <class InputIt>
+            ::std::vector<wchar_t> get_multistring_from_wstrings(const InputIt& first, const InputIt& last)

```

Comment:
As a note, `std::wstring` is allowed to contain embedded null characters. If it's easier for you, you can use std::wstring here instead of vector and then converting later from vector to wstring.

Reply:
when we had conversations with Stephan (STL), and looked up std::wstring, std::wstring is not allowed to have embedded nulls; that breaks its interface (not all methods work correctly) ... it's a somewhat common bug that some folks have when directly writing to a pre-allocated wstring to not trim the null character off the end.

Reply:
What does the standard say is the moral equivalent of PCWZZSTR? std::vector<std::wstring>>, right?

Reply:
wstring cannot have embedded nulls. I don't know of a moral equivalent that says "this has characters in it like a string. but also has N number of nulls - some in the middle - and multiple at the end". since the returned vector<wchar_t> must include all embedded nulls and all nulls (should be 2) at the end of the buffer.

Reply:
To wrap this discussion, the easiest STL way to represent "a sequence of chars that may include nulls" is `std::vector<wchar_t>`. Not very easy to do string-like stuff with, but that's the best we have for now.